### ISSUE TITLE
Can I link a module dynamically?

### ISSUE THREAD

**Issue by ales-tsurko** [koto-lang/koto#39]
I found how to compile the modules with Koto itself. But can I write a module in rust, compile it separately and then import it in already compiled Koto?

**Comment by irh** [comment]
There isn't a good solution for this yet. What you can do is have dynamic Rust libraries containing Koto modules that you load in your application, and then add the modules manually to the runtime's prelude. The trouble is that a lack of a stable ABI for Rust means that you have to be sure that the libraries and the application have been built with the same compiler version.

My current thinking is that a C API is needed for Koto, so that Koto modules can be built in Rust (or any other language that can work with a C API) without having to worry about the compiler version.

**Comment by ales-tsurko** [comment]
What do you think about a very little FFI, like just a function `make_module`, which returns some C struct with module info (name?) and a `ValueMap`? This won't make it possible to write modules on other languages yet, but at least using C API for this would allow to be sure in the ABI. This shouldn't take much work, I could provide a PR.

**Comment by irh** [comment]
I'm not an expert here but I think this won't be sufficient, because the layout of `ValueMap` isn't guaranteed to be stable between compiler versions. Something like a `make_map` / `insert_map_value` style of API would be needed, with the host application returning allocated structures and performing the data modifications (I think).

**Comment by ales-tsurko** [comment]
> the layout of ValueMap isn't guaranteed to be stable between compiler versions. 

I meant using not Rust, but C layout for that ([`repr(C)`](https://doc.rust-lang.org/nomicon/other-reprs.html#reprc)). Maybe I'm missing something, but as I see it, when we want to write a DLL, which is compatible with C, we just use `repr(C)` for types and `extern "C"` block to expose types and functions as FFI. And it's independent on Rust compiler's version while it's `cdylib`. And it works the other way around - we can use symbols from C DLL's in Rust using `repr(C)` and `extern` block. So we can just use C API for function `make_module` and make `ValueMap` `repr(c)` (or an opaque pointer?), so it'll work with C, but we actually use it in Rust.

**Comment by ales-tsurko** [comment]
Ah, just checked `ValueMap` and it looks to complex for that. So now I'm not sure in the above 😄

**Comment by irh** [comment]
I think you can add `repr(C)` to `ValueMap`, but there's more to the story than just representation, there's no guarantee that (e.g.) `HashMap` will have the same members between library versions. Also you would need this guaranteed layout for any `Value` that can be added to the module.

I think for this to work an API that's completely abstracted from specific data structrures is needed. I do share the desire for a more straight-forward option!

**Comment by ales-tsurko** [comment]
> I think you can add repr(C) to ValueMap, but there's more to the story than just representation, there's no guarantee that (e.g.) HashMap will have the same members between library versions. Also you would need this guaranteed layout for any Value that can be added to the module.

Yeah, when I was writing it I didn't know how complex is `ValueMap`.

> I think for this to work an API that's completely abstracted from specific data structrures is needed. I do share the desire for a more straight-forward option!

I'll look into it. I need to be more familiar with the code base to propose something.

**Comment by ales-tsurko** [comment]
So far I've found that it's actually bigger than I thought. For example, we can solve the issue with instantiating `ValueMap` as a module itself by allocating it on the Koto's side and passing it as an opaque pointer via some FFI. It's also easy to provide FFI functions to work with it, so we, for example, wouldn't need to cast it on the plugin's side. But still, we need to send/receive data somehow, so we need FFI for `Value` as well. Where we again need FFI for `ValueMap`, for example, and this time we need to be able to allocate it on the plugin's side. And as it's not `repr(c)` layout, we have the same issue we started with.

**Comment by irh** [comment]
Yes, to be completely clear I think a full C API for creation and modification of any `Value` types will be needed.

I'm curious about how you're planning to use dynamic loading? I had in mind that this would be important for package management but it hadn't occurred to me that it would be necessary within the context of an application.

**Comment by ales-tsurko** [comment]
I'd like to make it a library. So it would be possible to use Koto on its own and import that library. Without dynamic linking the only option I have is providing my own version of Koto with this library included, that is a little cumbersome. I have plans to make a little DAW-like application, where everything managed by scripting (= "live coding" DAW). But I was going to make it a separate project.

**Comment by irh** [comment]
Ah yes of course, this is a good thing to work towards.

My plan for the libraries in the [libs](https://github.com/koto-lang/koto/tree/6d0cf645cb0fa6f287c4af3d51e6ce96ddb6dde1/libs) folder is that once dynamic loading / package management is in place these libraries could be extracted from the repo. For now they get [explicitly added](https://github.com/koto-lang/koto/blob/6d0cf645cb0fa6f287c4af3d51e6ce96ddb6dde1/src/cli/src/main.rs#L116) when the Koto application is launched, which is non-ideal.

So I guess to summarize I would say that dynamic loading is something that I would love to see in Koto, but it (probably) requires a C API, and I've been deferring that effort until I'm more confident that the fundamentals of the language are stable. It's achievable, just not at the top of the todo list at the moment.

**Comment by ales-tsurko** [comment]
There are other options. First, there's a crate [abi_stable](https://docs.rs/abi_stable/0.9.3/abi_stable/). But it'll introduce a new dependency and will allow Rust only libraries. Another option could be an integrated library for FFI, to allow calling C functions and initializing C structures as maps in Koto. This looks like less work to do than providing a full C API. I could look into it after the weekend.

**Comment by ales-tsurko** [comment]
So, yes - an FFI library for Koto could be an option. Something like [node-ffi](https://github.com/node-ffi/node-ffi/wiki/Node-FFI-Tutorial) would work. But it would still take some time to implement, so I'll go with my own distribution of Koto for now and see how it goes. I'd like to return to this when I have more time. Still I'm ready to help with #36 if needed.

**Comment by ales-tsurko** [comment]
I now think that it would be great to use some wasm runtime (wasmer, for example) and support libraries compiled into wasm. This would allow writing libraries in other languages and extend the infrastructure without recompiling the project. It also would be great to be able to compile koto into wasm.

**Comment by irh** [comment]
Hey @ales-tsurko 👋

I'm interested in using wasm as the primary extension runtime too, I like the idea of a library of platform-independent precompiled modules that can be easily pulled into your system without having to worry about local compilation. 

I suspect that native dynamic libraries would also be wished for to allow for maximum performance, so I think ideally the 'C API' that I mentioned above would really be an 'FFI API'. I don't have a clear idea how to do that though, one of these days I'll get a chance to sketch out a prototype.

> It also would be great to be able to compile koto into wasm.

This already works! Take a look in the `examples/wasm` folder.

**Comment by irh** [comment]
Oh interesting I see.. Something to keep in mind for the wasm-y future =)

**Comment by ales-tsurko** [comment]
If you would accept a PR with implementation for wasm modules importing, I could look into this when I have free time. It requires embedding a wasm runtime.
